desktop/src/preferences/write.rs

```diff
@@ -51,127 +51,251 @@ impl<'a> PreferencesWriter<'a> {
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use crate::preferences::read::read_preferences;
-    use fluent_templates::loader::langid;
+pub struct BookmarksWriter<'a>(&'a mut BookmarksAndDocument);
 
-    fn parse(input: &str) -> PreferencesAndDocument {
-        let (result, document) = read_preferences(input);
-        PreferencesAndDocument {
-            toml_document: document,
-            values: result.result,
-        }
+impl<'a> BookmarksWriter<'a> {
+    pub(super) fn new(bookmarks: &'a mut BookmarksAndDocument) -> Self {
+        Self(bookmarks)
     }
 
-    fn check_roundtrip(preferences: &mut PreferencesAndDocument) {
-        let read_result = read_preferences(&preferences.toml_document.to_string());
-        assert_eq!(
-            preferences.values, read_result.0.result,
-            "roundtrip failed: expected != actual"
-        );
+    fn bookmark_table(&mut self, index: usize) -> &mut Table {
+        self.get_underlying_table()
+            .get_mut(index)
+            .expect("invalid bookmark index")
     }
 
-    fn test(original: &str, fun: impl FnOnce(&mut PreferencesWriter), expected: &str) {
-        let mut preferences = parse(original);
-        let mut writer = PreferencesWriter::new(&mut preferences);
-        fun(&mut writer);
-        check_roundtrip(&mut preferences);
-        assert_eq!(expected, preferences.toml_document.to_string());
-    }
+    fn get_underlying_table(&mut self) -> &mut ArrayOfTables {
+        if self.0.toml_document.contains_array_of_tables("bookmark") {
+            return self.0.toml_document["bookmark"]
+                .as_array_of_tables_mut()
+                .expect("type was just verified");
+        }
 
-    #[test]
-    fn set_graphics_backend() {
-        test(
-            "",
-            |writer| writer.set_graphics_backend(GraphicsBackend::Default),
-            "graphics_backend = \"default\"\n",
-        );
+        // Type was incorrect, time to destroy.
+        self.0.toml_document.insert("bookmark", array());
+        self.0.toml_document["bookmark"]
+            .as_array_of_tables_mut()
+            .expect("type was just created")
+    }
 
-        test(
-            "graphics_backend = \"fast\"",
-            |writer| writer.set_graphics_backend(GraphicsBackend::Vulkan),
-            "graphics_backend = \"vulkan\"\n",
-        );
+    pub fn add(&mut self, bookmark: Bookmark) {
+        let table = self.get_underlying_table();
+        // TODO: If we add a BookmarkWriter use this here instead rather than duplicating the table write code.
```

Comment:
Isn't this struct `BookmarkWriter`? Am I misunderstanding this comment?

Reply:
Forgot to remove this TODO, sorry.